src/Language/GraphQL/Draft/Syntax.hs

```diff
@@ -115,15 +124,47 @@ newtype Name = Name {unName :: Text}
   deriving stock (Eq, Lift, Ord, Show)
   deriving newtype (Hashable, NFData, Pretty, Semigroup, J.ToJSONKey, J.ToJSON)
 
+-- | @NameSuffix@ is essentially a GQL identifier that can be used as Suffix
+--  It is slightely different from @Name@ as it relaxes the criteria that a
+--  @Name@ cannot start with a digit.
+type NameSuffix :: Type
+newtype NameSuffix = Suffix {unNameSuffix :: Text}
+  deriving stock (Lift, Show)
+
+matchFirst :: Char -> Bool
+matchFirst c = c == '_' || C.isAsciiUpper c || C.isAsciiLower c
+
+matchBody :: Char -> Bool
+matchBody c = c == '_' || C.isAsciiUpper c || C.isAsciiLower c || C.isDigit c
+
+isValidName :: Text -> Bool
+isValidName text =
+  case T.uncons text of
+    Nothing -> False
+    Just (first, body) ->
+      matchFirst first && T.all matchBody body
+
 mkName :: Text -> Maybe Name
 mkName text =
-  T.uncons text >>= \(first, body) ->
-    if matchFirst first && T.all matchBody body
-      then Just (Name text)
-      else Nothing
+  if isValidName text
+    then Just (Name text)
+    else Nothing
+
+mkNameSuffix :: Text -> Maybe NameSuffix
+mkNameSuffix text =
+  if T.all matchBody text
+    then Just (Suffix text)
+    else Nothing
+
+addSuffixes :: Name -> [NameSuffix] -> Name
+addSuffixes prefix [] = prefix
+addSuffixes (Name prefix) suffs = Name $ T.concat (prefix : suffsT)
   where
-    matchFirst c = c == '_' || C.isAsciiUpper c || C.isAsciiLower c
-    matchBody c = c == '_' || C.isAsciiUpper c || C.isAsciiLower c || C.isDigit c
+    suffsT = map unNameSuffix suffs
```

Comment:
i think this is just `coerce`, but it might not work without an explicit type annotation.